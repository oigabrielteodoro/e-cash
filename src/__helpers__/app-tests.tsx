import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
