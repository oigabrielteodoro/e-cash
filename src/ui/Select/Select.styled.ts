import styled, { css } from 'styled-components'
import { motion } from 'framer-motion'

import { theme } from 'config'
import { Container as InputContainer } from 'ui/Input/Input.styled'
import { fadeIn } from 'ui/_animations'

type Props = {
  isOpen: boolean
}

export const Container = styled(InputContainer)<Props>`
  ${({ isOpen }) =>
    isOpen &&
    css`
      border-radius: 0.5rem 0.5rem 0 0;
      border-color: ${theme.colors.neutral[300]};
      box-shadow: none;

      &:hover {
        border-color: ${theme.colors.neutral[300]};
      }

      svg {
        transform: rotate(-180deg);
      }
    `}
`

export const Dropdown = styled(motion.ul).attrs(fadeIn())`
  position: absolute;
  background: ${theme.colors.white};
  width: 100%;
  top: calc(100% - 0.1rem);
  border: 0.1rem solid ${theme.colors.neutral[300]};
  border-radius: 0 0 ${theme.radius.popover} ${theme.radius.popover};
`
