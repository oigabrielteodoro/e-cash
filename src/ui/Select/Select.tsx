import React, {
  useRef,
  useState,
  useEffect,
  cloneElement,
  ReactElement,
  ChangeEvent,
} from 'react'
import { AiOutlineExclamationCircle } from 'react-icons/ai'
import { FiChevronDown } from 'react-icons/fi'
import { AnimatePresence } from 'framer-motion'
import { useFormContext } from 'react-hook-form'

import toArray from 'lodash/toArray'

import { ClickOutsideElement, InputProps } from 'ui'
import { useInput } from 'hooks'

import * as InputStyled from 'ui/Input/Input.styled'

import { Option, OptionProps } from './Option'

import * as S from './Select.styled'

type SelectProps = {
  name: string
  children: ReactElement<OptionProps>[]
} & Omit<InputProps, 'icon'>

export function Select({
  id,
  name,
  label,
  variant = 'primary',
  defaultValue,
  error,
  children,
  onBlur,
  onFocus,
  ...rest
}: SelectProps) {
  const containerRef = useRef<HTMLDivElement>(null)
  const inputRef = useRef<HTMLInputElement>(null)

  const {
    isFilled,
    isFocused,
    isErrored,
    errorMessage,
    handleOnBlur,
    handleOnFocus,
  } = useInput({
    defaultValue,
    error,
    onBlur,
    onFocus,
  })

  const { register, setValue } = useFormContext()

  const [isOpen, setIsOpen] = useState(false)
  const [filteredOptions, setFilteredOptions] = useState(toArray(children))

  useEffect(() => {
    register(name)
  }, [name, register])

  useEffect(() => {
    if (defaultValue && inputRef?.current) {
      const allOptions = toArray(children)
      const foundOption = allOptions.find(
        (option) => option.props.value === defaultValue,
      )

      if (foundOption) {
        inputRef.current.value = foundOption.props.children
      }
    }
  }, [defaultValue, children])

  function handleOnClick(value = '', children = '') {
    setValue(name, value)

    if (inputRef?.current) {
      inputRef.current.value = children
    }

    setIsOpen(false)
  }

  function handleOnChange(event: ChangeEvent<HTMLInputElement>) {
    const allOptions = toArray(children)
    const inputValue = event.currentTarget.value

    if (inputValue?.length === 0) {
      setValue(name, '')
    }

    setFilteredOptions(
      allOptions.filter((child) =>
        child.props.children.toLowerCase().includes(inputValue.toLowerCase()),
      ),
    )
  }

  return (
    <S.Wrapper>
      <InputStyled.Wrapper as='div'>
        <label htmlFor={id}>{label}</label>
        <S.Container
          ref={containerRef}
          variant={variant}
          isOpen={isOpen}
          isFilled={isFilled}
          isFocused={isFocused}
          isErrored={isErrored}
          onClick={() => setIsOpen(true)}
        >
          <input
            id={id}
            ref={inputRef}
            name={name}
            aria-label={name}
            defaultValue={defaultValue}
            onBlur={handleOnBlur}
            onFocus={handleOnFocus}
            onChange={handleOnChange}
            autoComplete='off'
            role='combobox'
            {...rest}
          />

          <FiChevronDown size={20} />
        </S.Container>
        {error && !isOpen && (
          <InputStyled.ErrorContainer>
            <AiOutlineExclamationCircle size={18} />
            <span>{errorMessage}</span>
          </InputStyled.ErrorContainer>
        )}
      </InputStyled.Wrapper>
      <AnimatePresence>
        {isOpen && (
          <ClickOutsideElement
            isOpen={isOpen}
            ignoredRefs={[containerRef]}
            onClose={() => setIsOpen(false)}
          >
            <S.Dropdown>
              {filteredOptions.map((child) => {
                const props = {
                  ...child.props,
                  key: `Select.Option.${child.props.value}`,
                  onClick: () =>
                    handleOnClick(child.props.value, child.props.children),
                }

                return cloneElement(child, props)
              })}
            </S.Dropdown>
          </ClickOutsideElement>
        )}
      </AnimatePresence>
    </S.Wrapper>
  )
}

Select.Option = Option
