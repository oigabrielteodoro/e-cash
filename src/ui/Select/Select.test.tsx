import React from 'react'
import { useForm, FormProvider } from 'react-hook-form'

import { render, screen, userEvent } from '__helpers__/app-tests'

import { Select } from './Select'

type Props = {
  error?: string
  defaultValue?: string
}

function MockedComponent({ error, defaultValue }: Props) {
  const form = useForm({
    defaultValues: {
      option: defaultValue,
    },
  })

  return (
    <FormProvider {...form}>
      <Select
        name='option'
        label='Option'
        error={error}
        defaultValue={defaultValue}
      >
        <Select.Option value='option_1'>Option 1</Select.Option>
        <Select.Option value='option_2'>Option 2</Select.Option>
        <Select.Option value='option_3'>Option 3</Select.Option>
      </Select>
    </FormProvider>
  )
}

describe('Select', () => {
  it('should be able match snapshot', () => {
    const { container } = render(<MockedComponent />)

    expect(container).toMatchSnapshot()
  })

  it('should be able render correctly', () => {
    render(<MockedComponent />)

    expect(screen.getByRole('combobox')).toBeInTheDocument()
  })

  it('should be able render with initial values', async () => {
    const valueToType = 'Option 1'

    render(<MockedComponent defaultValue={valueToType} />)

    expect(await screen.findByRole('combobox')).toHaveValue(valueToType)
  })

  it('should be able open/close successfully', async () => {
    render(<MockedComponent />)

    const input = screen.getByRole('combobox')

    expect(input).toBeInTheDocument()

    userEvent.click(input)

    expect(
      await screen.findByRole('option', {
        name: 'Option 1',
      }),
    ).toBeInTheDocument()
    expect(
      await screen.findByRole('option', {
        name: 'Option 2',
      }),
    ).toBeInTheDocument()
    expect(
      await screen.findByRole('option', {
        name: 'Option 3',
      }),
    ).toBeInTheDocument()
  })

  it('should be able show options based on what user has typed', async () => {
    render(<MockedComponent />)

    const input = screen.getByRole('combobox')

    expect(input).toBeInTheDocument()

    userEvent.click(input)
    userEvent.type(input, 'Option 1')

    expect(
      await screen.findByRole('option', {
        name: 'Option 1',
      }),
    ).toBeInTheDocument()
    expect(
      screen.queryByRole('option', {
        name: 'Option 2',
      }),
    ).not.toBeInTheDocument()
    expect(
      screen.queryByRole('option', {
        name: 'Option 3',
      }),
    ).not.toBeInTheDocument()
  })

  it('should be able render error message', async () => {
    render(<MockedComponent error='Value is required' />)

    expect(screen.getByRole('combobox')).toBeInTheDocument()
    expect(screen.getByText('Value is required')).toBeInTheDocument()
  })
})
