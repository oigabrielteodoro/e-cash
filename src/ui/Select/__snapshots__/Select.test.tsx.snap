// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should be able match snapshot 1`] = `
.c1 {
  border: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  position: relative;
}

.c1 label {
  position: absolute;
  font-size: 1.4rem;
  left: 1.2rem;
  top: -0.1rem;
  background: #ffffff;
  padding: 0 0.8rem;
  color: #a1a7ae;
  font-weight: 500;
  cursor: pointer;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0.1rem solid #a1a7ae;
  border-radius: 0.5rem;
  margin-top: 0.8rem;
  -webkit-transition: 300ms;
  transition: 300ms;
  width: 100%;
  background: #ffffff;
}

.c2 svg {
  margin-left: auto;
  margin-right: 2rem;
  color: #a1a7ae;
  -webkit-transition: 300ms;
  transition: 300ms;
}

.c2:hover {
  border: 0.1rem solid #3a67e1;
}

.c2:hover svg {
  color: #3a67e1;
}

.c2 input {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: 0;
  background: transparent;
  padding: 1.8rem 2rem;
  color: #18171d;
  font-size: 1.6rem;
  border-radius: 0.5rem;
}

.c2 input::-webkit-input-placeholder {
  color: #a1a7ae;
}

.c2 input::-moz-placeholder {
  color: #a1a7ae;
}

.c2 input:-ms-input-placeholder {
  color: #a1a7ae;
}

.c2 input::placeholder {
  color: #a1a7ae;
}

.c0 {
  border: 0;
  position: relative;
}

<div>
  <fieldset
    class="c0"
  >
    <div
      class="c1"
    >
      <label>
        Option
      </label>
      <div
        class="Selectstyled__Container-sc-1fbxwyn-1 c2"
      >
        <input
          aria-label="option"
          autocomplete="off"
          name="option"
          role="combobox"
          value=""
        />
        <svg
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
    </div>
  </fieldset>
  <div
    class="Toastify"
  />
</div>
`;
